### ISSUE TITLE
v6: Change in subdocument not tracked using Object.assign / spread operator

### ISSUE THREAD

**Issue by saveman71** [Automattic/mongoose#10773]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

In some cases, change tracking in nested objects doesn't work.

**If the current behavior is a bug, please provide the steps to reproduce.**

```js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const PaymentSchema = new Schema({
  status: String,
}, {
  _id: false,
});

const OrderSchema = new Schema({
  status: String,
  payments: {
    payout: PaymentSchema,
  },
})

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test-repro');
  await mongoose.connection.dropDatabase();
  const Order = mongoose.model('Order', OrderSchema);

  const order = new Order({
    status: 'unpaid',
    payments: {
      payout: {
        status: 'unpaid',
      },
    }
  });

  await order.save();

  // It works with:
  // const newPaymentsStatus = {payout: {status: 'unpaid'}};

  // Object.assign or spread operator will break it
  const newPaymentsStatus = Object.assign({}, order.payments);

  order.status = 'paid';
  newPaymentsStatus.payout.status = 'paid';

  order.payments = newPaymentsStatus;
  console.log(order.isModified('payments'));

  await order.save().then(order => {
    // Pretends that the new object has the correct info
    console.log(order);
  });

  const notUpdated = await Order.findById({_id: order._id});

  console.log(notUpdated);
}

run();
```

**What is the expected behavior?**

Change is registered like it used to be in mongoose 5.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

MongoDB 3.4
Mongoose 6.0.6
NodeJS v14.17.6

**Comment by saveman71** [comment]
So, as it used to be working in mongoose 5, I'm submitting this as a bug, but maybe this is too difficult to fix / is expected behaviour.

However, it's not so hard to end up with this kind of code and it's definitely unexpected behavior.

**Comment by vkarpov15** [comment]
Thanks for reporting this issue and providing a detailed repro. This is an unfortunate bug in how Mongoose tracks saved state, and is related to our work on #9396 for 5.11 and #10596 for 6.0.

The idea of _saved state_ is to avoid unnecessary saves when you set a property to a different value, and then back to the initial value.

```javascript
const doc = await Model.findOne();

const originalValue = doc.myStr;
doc.myStr = originalValue + ' and some more characters';
doc.myStr = originalValue;

await doc.save(); // <-- Should not save `myStr` because the value was changed back to the initial value
```

But this gets a bit tricky with nested objects. `doc.nested.myStr = originalValue + ' and some more characters'` and then `doc.nested = { myStr: originalValue }` is a bit trickier because you need to know that `nested`'s saved value is given the original value for `nested.myStr`.

The way we fixed this in c976b99 is just pre-emptively storing all of `doc.nested` in `savedState` when `doc.nested.myStr` is set. This isn't optimal in terms of memory usage, but more durable because there's less likely to be bugs with this approach.

**Comment by saveman71** [comment]
Thanks a lot for investigating, the explanation, and for the fix!
